src/modules/zfs/ZfsJob.cpp

```diff
@@ -0,0 +1,326 @@
+/* === This file is part of Calamares - <https://calamares.io> ===
+ *
+ *   SPDX-FileCopyrightText: 2021 Evan James <dalto@fastmail.com>
+ *   SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ *   Calamares is Free Software: see the License-Identifier above.
+ *
+ */
+
+#include "ZfsJob.h"
+
+#include "utils/CalamaresUtilsSystem.h"
+#include "utils/Logger.h"
+#include "utils/Variant.h"
+
+#include "GlobalStorage.h"
+#include "JobQueue.h"
+#include "Settings.h"
+
+#include <QProcess>
+
+ZfsJob::ZfsJob( QObject* parent )
+    : Calamares::CppJob( parent )
+{
+}
+
+ZfsJob::~ZfsJob() {}
+
+QString
+ZfsJob::prettyName() const
+{
+    return tr( "Create ZFS pools and datasets" );
+}
+
+QString
+ZfsJob::AlphaNumeric( QString input ) const
+{
+    return input.remove( QRegExp( "[^a-zA-Z\\d\\s]" ) );
+}
+
+void
+ZfsJob::CollectMountpoints( const QVariantList& partitions )
+{
+    m_mountpoints.empty();
+    for ( const QVariant& partition : partitions )
+    {
+        if ( partition.canConvert( QVariant::Map ) )
+        {
+            QString mountpoint = partition.toMap().value( "mountPoint" ).toString();
+            if ( !mountpoint.isEmpty() )
+            {
+                m_mountpoints.append( mountpoint );
+            }
+        }
+    }
+}
+
+bool
+ZfsJob::IsMountpointOverlapping( const QString& targetMountpoint ) const
+{
+    for ( const QString& mountpoint : m_mountpoints )
+    {
+        if ( mountpoint != '/' && targetMountpoint.startsWith( mountpoint ) )
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+
+ZfsResult
+ZfsJob::CreateZpool( QString deviceName, QString poolName, QString poolOptions, bool encrypt, QString passphrase ) const
+{
+    // zfs doesn't wait for the devices so pause for 2 seconds to ensure we give time for the device files to be created
+    QString command;
+    if ( encrypt )
+    {
+        command = "sleep 2 ; echo \"" + passphrase + "\" | zpool create " + poolOptions
+            + " -O encryption=aes-256-gcm -O keyformat=passphrase " + poolName + " " + deviceName;
+    }
+    else
+    {
+        command = "sleep 2 ; zpool create " + poolOptions + " " + poolName + " " + deviceName;
+    }
+
+    // We use a qProcess instead of runCommand so the password will not end up in the logs
```

Comment:
If the password is ending up in the logs, then `RedactedCommand` needs some more work; but I think that using `Runner` or the simplified API and setting stdin is a better approach. (Discussion welcome -- I really don't like commands that do not end up in the logs, but that **can** cause errors, because they're hard to debug)